lib/Sema/Comment.cpp

```diff
@@ -0,0 +1,224 @@
+    Type vdComparisonTy = VD->getInterfaceType();
+    if (!vdComparisonTy) {
+      return std::nullopt;
+    }
+    if (auto fnTy = vdComparisonTy->getAs<AnyFunctionType>()) {
+      // Strip off the 'Self' parameter.
+      vdComparisonTy = fnTy->getResult();
+    }
+
+    for (auto *member : members) {
+      if (VD == member || member->isInvalid()) {
+        continue;
+      }
+      if (!isa<TypeDecl>(member)) {
+        if (!swift::TypeChecker::witnessStructureMatches(member, VD)) {
+          continue;
+        }
+        Type memberComparisonTy = member->getInterfaceType();
+        if (!memberComparisonTy) {
+          continue;
+        }
+        if (auto fnTy = memberComparisonTy->getAs<AnyFunctionType>()) {
+          // Strip off the 'Self' parameter.
+          memberComparisonTy = fnTy->getResult();
+        }
+        if (!vdComparisonTy->matches(memberComparisonTy,
+                                     TypeMatchFlags::AllowOverride)) {
+          continue;
+        }
+      }
+      auto newResult = visit(member);
+      if (!newResult)
+        continue;
+
+      if (result) {
+        // Found two or more decls with doc-comment.
+        return std::nullopt;
+      }
+      result = newResult;
+    }
+    return result;
+  }
+
+  std::optional<ResultWithDecl> findRequirementDecl(const ValueDecl *VD) {
+    std::queue<const ValueDecl *> requirements;
+    while (true) {
+      for (auto *req : VD->getSatisfiedProtocolRequirements()) {
+        if (auto result = visit(req))
+          return result;
+
+        requirements.push(req);
+      }
+      if (requirements.empty())
+        return std::nullopt;
+
+      VD = requirements.front();
+      requirements.pop();
+    }
+  }
+
+public:
+  std::optional<ResultWithDecl> findCommentProvider(const Decl *D) {
+    if (auto result = visit(D))
+      return result;
+
+    auto *VD = dyn_cast<ValueDecl>(D);
+    if (!VD)
+      return std::nullopt;
+
+    if (auto result = findOverriddenDecl(VD))
+      return result;
+
+    if (auto result = findRequirementDecl(VD))
+      return result;
+
+    if (auto result = findDefaultProvidedDecl(VD))
+      return result;
+
+    return std::nullopt;
+  }
+};
+} // end anonymous namespace
+
+const Decl *Decl::getDocCommentProvidingDecl() const {
+  // Search for the first decl we see with a non-empty raw comment.
+  auto finder = CommentProviderFinder<RawComment>(
+      [](const Decl *D) -> std::optional<RawComment> {
+        auto comment = D->getRawComment();
+        if (comment.isEmpty())
+          return std::nullopt;
+        return comment;
+      });
+
+  auto result = finder.findCommentProvider(this);
+  return result ? result->second : nullptr;
+}
+
+StringRef swift::Decl::getSemanticBriefComment() const {
+  if (!canHaveComment())
+    return StringRef();
+
+  auto &eval = getASTContext().evaluator;
+  return evaluateOrDefault(eval, SemanticBriefCommentRequest{this},
+                           StringRef());
+}
```

Comment:
I probably should have clarified why requests are able to do layering violations like this, it's because the request classes themselves are in libAST, but their evaluation functions are registered at runtime

Reply:
Ah, I see. That makes me less concerned about the layering as well. I created requests fro both `getSemanticBriefComment` and `getDocCommentProvidingDecl` and moved the code that calls them to `libAST`.